user/tests/integration/ota/factory_reset/factory_reset.spec.js

```diff
@@ -0,0 +1,127 @@
+suite('Factory Reset');
+
+platform('gen3');
+systemThread('enabled');
+
+const { HalModuleParser, ModuleInfo, updateModulePrefix, updateModuleSuffix, updateModuleCrc32 } = require('binary-version-reader');
+const tempy = require('tempy');
+
+const { readFile } = require('fs').promises;
+const path = require('path');
+
+let api = null;
+let auth = null;
+let device = null;
+let deviceId = null;
+
+async function flash(ctx, binFile, { timeout = 120000, mayFail = false } = {}) {
+	const appName = path.basename(binFile, '.bin');
+	await api.flashDevice({ deviceId, files: { [appName]: binFile }, auth });
+	const ok = await waitFlashStatusEvent(ctx, timeout);
+	if (!ok && !mayFail) {
+		throw new Error('Update failed');
+	}
+	return ok;
+}
+
+async function waitFlashStatusEvent(ctx, timeout) {
+	let timeoutAt = Date.now() + timeout;
+	let data = null;
+	for (;;) {
+		const t = timeoutAt - Date.now();
+		if (t <= 0) {
+			throw new Error("Event timeout");
+		}
+		data = await ctx.particle.receiveEvent('spark/flash/status', { timeout: t });
+		ctx.particle.log.verbose('spark/flash/status:', data);
+		if (data.startsWith('success')) {
+			return true;
+		}
+		if (data.startsWith('failed')) {
+			return false;
+		}
+	}
+}
+
+async function delay(ms) {
+	return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+before(function() {
+	api = this.particle.apiClient.instance;
+	auth = this.particle.apiClient.token;
+	device = this.particle.devices[0];
+	deviceId = device.id;
+});
+
+test('01_disable_resets_and_connect', async function () {
+	// See the test app
+});
+
+test('02_flash_binary', async function () {
```

Comment:
There should be an easy way to refactor this to a single function, I had a hard time debugging it so left it as is for now